analyses/long-format-table-utils/annotator/download-annotation-data.R

```diff
@@ -0,0 +1,205 @@
+# Download gene Ensembl ENSG ID to gene full name and protein RefSeq IDs mapping
+# file `annotation-data/ensg-gene-full-name-refseq-protein.tsv` from
+# https://mygene.info/
+
+# Import functions -------------------------------------------------------------
+# Get %>% without loading the whole library
+`%>%` <- dplyr::`%>%`
+
+
+
+# Define functions -------------------------------------------------------------
+# Collapse a list of refseq.protein character vectors from mygene.info query
+# results
+#
+# Args:
+# - rp_vec_list: list of refseq.protein character vectors from mygene.info query
+#   results
+#
+# Returns a single character value of a comma-separated refseq.protein value or
+# NA
+collapse_rp_lists <- function(rp_vec_list) {
+  # remove list elements that are NULL
+  rm_null_rp_vec_list <- purrr::discard(rp_vec_list, is.null)
+  # assert non-null list elements are all characters
+  lapply(rm_null_rp_vec_list, function(x) {
+    if (!is.character(x)) {
+      stop(paste0("mygene query returns non-character refseq.protein.\n",
+                  "Check query results. Revise download-annotation-data.R ",
+                  "to handle non-character values."))
+    }
+  })
+
+  # combimed vector of unique refseq.protein values
+  uniq_c_rm_null_rp_vec <- sort(unique(
+    purrr::reduce(rm_null_rp_vec_list, c, .init = character(0))))
+  # remove NA
+  rm_na_uniq_c_rm_null_rp_vec <- purrr::discard(
+    uniq_c_rm_null_rp_vec, is.na)
+  # keep only NP_### refseq.protein values
+  np_rp_vec <- purrr::keep(
+    rm_na_uniq_c_rm_null_rp_vec, function(x) stringr::str_detect(x, "^NP_"))
+
+  clp_np_rp_str <- paste(np_rp_vec, collapse = ",")
+
+  if (identical(clp_np_rp_str, "")) {
+    # no NP_### value for the gene query
+    # dplyr::summarise needs function return value to be character for a
+    # character column. as.character(NA) is still NA rather than "NA".
+    return(as.character(NA))
+  } else{
+    return(clp_np_rp_str)
+  }
+}
+# # test cases
+# collapse_rp_lists(list(NULL, NULL, NULL))
+# collapse_rp_lists(list())
+# collapse_rp_lists(list(c("NP_1", "NP_2"), c("NP_1", "NP_3")))
+# collapse_rp_lists(list(c("NX_1", "NA"), c("NP_1", "NP_3")))
+# collapse_rp_lists(list(c("NP_1", "NP_2"), NULL, c("NP_1", "NP_3")))
+# collapse_rp_lists(list(c("NP_3", "NP_2")))
+# collapse_rp_lists(list(c("NP_1", "NP_2"), NULL, c("NP_1", "NP_3", NA),
+#                        c(NA, NA, character(0)), c(NA, character(0)),
+#                        character(0)))
```

Comment:
Will this and the next commented "test cases" be removed at a later time?

Reply:
I prefer to leave them there, as these lines could be helpful to future updates if mygene.info API changes, and we currently do not have standardized support for unit testing.

We could definitely remove them here, and store them somewhere else. We could also discuss adding a standardized support for unit testing.

Reply:
sure, that's fine. looping in @yuankunzhu for thoughts on adding unit testing support.

Reply:
I have implemented a lightweight unit testing framework for the `annotator` submodule using the [`testthat`](https://testthat.r-lib.org/index.html) package. The `testthat` package is available in the Docker image.

To run all tests, run `bash analyses/long-format-table-utils/annotator/run-tests.sh` from any working directory. Following is an example run.

```text
~/OpenPedCan-analysis$ bash analyses/long-format-table-utils/annotator/run-tests.sh
✔ |  OK F W S | Context
✔ |   8       | tests/test_collapse_name_vec.R
✔ |   7       | tests/test_collapse_rp_lists.R
✔ |   5       | tests/test_collapse_rp_lists.R

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 0.2 s

OK:       20
Failed:   0
Warnings: 0
Skipped:  0
Done running run-tests.sh
```

This is implemented in the <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/55/commits/48a131dbbd87c5cfc13bd52a21035763147c4c76> commit.

Reply:
Thanks for the nifty tests, they all passed for me.